### DISCUSSION TITLE
关于事件冷却时间的讨论

### DISCUSSION THREAD

**Discussion by jks15satoshi** [nonebot/discussions#2] [通用]
在翻阅现有 Nonebot 2 的 API 文档后，我并没有发现有关事件触发冷却时间的调用接口（应该是没有，如果有的话麻烦骂醒我）。

那么如果要自己手动实现事件冷却时间判定，有没有一个比较好的解决方案？我自己的想法是在每次事件被触发时在 `Matcher.finish()` 执行前记录当前时间，通过自定义 `Rule` 判断是否在 CD 时间范围内，但感觉有点麻烦。

另外未来 Nonebot 会不会考虑加入这个特性？

**Comment by stdrc** [comment]
意思是用户调用频率限制？

**Reply by synodriver** [reply]
建议以插件形式提供，Rule可以做到

**Reply by stdrc** [reply]
此处可以 @yanyongyu

**Reply by jks15satoshi** [reply]
是这个意思，我的想法是写成一个通用的方法，可以为不同的事件设置不同的周期频率

**Reply by yanyongyu** [reply]
如果由nonebot来实现可以实现matcher的执行限制，但没法做到为每一个不同的用户触发进行限制

**Reply by synodriver** [reply]
每一个用户都要不同限制的话得存文件吧

**Reply by jks15satoshi** [reply]
> 建议以插件形式提供，Rule可以做到

其实我也是想问问有没有什么更合适的方案，关于判定 CD 这方面我也是想着用 `Rule` 去实现，但写下来总觉得有些繁琐（其实就是太菜

**Reply by synodriver** [reply]
> > 建议以插件形式提供，Rule可以做到
> 
> 其实我也是想问问有没有什么更合适的方案，关于判定 CD 这方面我也是想着用 `Rule` 去实现，但写下来总觉得有些繁琐（其实就是太菜

可以用redis实现，set(user_id,1,ex=过期时间) ， 然后检测redis，有这个key就说明一定时间内调用过了不能再调用了

**Reply by jks15satoshi** [reply]
> > > 建议以插件形式提供，Rule可以做到
> > 
> > 
> > 其实我也是想问问有没有什么更合适的方案，关于判定 CD 这方面我也是想着用 `Rule` 去实现，但写下来总觉得有些繁琐（其实就是太菜
> 
> 可以用redis实现，set(user_id,1,ex=过期时间) ， 然后检测redis，有这个key就说明一定时间内调用过了不能再调用了

好的

**Reply by mnixry** [reply]
> 每一个用户都要不同限制的话得存文件吧

没有必要持久化,存储在内存里面就可以
用Redis还要外置一个服务,太重了


我的想法是直接整个全局dict变量存`qq号:上一事件时间戳`
下一次触发该事件的时候查一下,和当前时间戳一减就行
